docs/design/validation.md

```diff
@@ -0,0 +1,2048 @@
+}
+
+func (v stringLengthBetweenValidator) Validate(ctx context.Context, path *tftypes.AttributePath, rawValue attr.Value) error {
+    value, ok := rawValue.(types.String)
+    
+    if !ok {
+        return fmt.Errorf("%s with incorrect type: %T", path, rawValue)
+    }
+
+    if value.Unknown {
+        return fmt.Errorf("%s with unknown value", path)
+    }
+
+    if len(value.Value) < v.minimum || len(value.Value) > v.maximum {
+        return fmt.Errorf("%s with value %q %s", path, value.Value, v.Description(ctx))
+    }
+
+    return nil
+}
+
+func StringLengthBetween(minimum int, maximum int) stringLengthBetweenValidator {
+    return stringLengthBetweenValidator{
+        maximum: maximum,
+        minimum: minimum,
+    }
+}
+```
+
+This helps solve the documentation issue with the following example slice type alias and receiver method:
+
+```go
+// ValueValidators implements iteration functions across ValueValidator
+type ValueValidators []ValueValidator
+
+// Descriptions returns all ValueValidator Description
+func (vs ValueValidators) Descriptions(ctx context.Context) []string {
+    result := make([]string, 0, len(vs))
+
+    for _, v := range vs {
+        result = append(result, v.Description(ctx))
+    }
+    return result
+}
+```
+
+To support passing through the provider instance, a separate interface type could be introduced that includes a function call with the `tfsdk.Provider` interface type:
+
+```go
+type ValueValidatorWithProvider interface {
+    ValueValidator
+    ValidateWithProvider(context.Context, provider tfsdk.Provider, path *tftypes.AttributePath, value attr.Value) error
+}
+```
+
+#### Attribute Value Validation Function Value Parameter
+
+Regardless the choice of concrete or interface types for the value validation functions, the parameters and returns for the implementations will play a crucial role on the extensibility and development experience.
+
+##### `attr.Value` Type
+
+The simplest implementation in the framework that could occur in all function types or interfaces is directly supplying an `attr.Value` and requiring implementations to handle all type conversion:
+
+```go
+func (v someValidator) Validate(ctx context.Context, path *tftypes.AttributePath, rawValue attr.Value) error {
+    value, ok := rawValue.(types.String)
+    
+    if !ok {
+        return fmt.Errorf("%s with incorrect type: %T", path, rawValue)
+    }
+
+    // ... rest of logic ...
+```
+
+This proposal would strongly encourage all implementations to handle value type conversions since using a stronger type in the function signature would risk panics that the framework cannot prevent and the compiler cannot check. Any error handling here could become inconsistent across implementations. This type conversion logic feels like an unnecessary burden on implementors and could reduce the developer experience as this logic would always need to be repeated with little to no actual utility.
```

Comment:
I apologize for my poor wording there. Given the scenario that a developer wrote a validation like the below:

```go
func (v someValidator) Validate(ctx context.Context, path *tftypes.AttributePath, value types.String) error {
    // ... rest of logic ...
```

But injected it into the wrong attribute type:

```go
schema.Attribute{
    Type: types.NumberType,
    Validators: /* validation func from above */,
}
```

I _think_ the Go compiler would be fine with it since `types.String` satisfies the interface, but it would panic at runtime? I guess I should try it and reword this better.

Reply:
Seems like the compiler catches this hypothetical, so I'll remove this.

```
./prog.go:99:22: cannot use StringLengthBetween(1, 5) (type stringLengthBetweenValidator) as type ValueValidator in slice literal:
	stringLengthBetweenValidator does not implement ValueValidator (wrong type for Validate method)
		have Validate(context.Context, *tftypes.AttributePath, types.String) error
		want Validate(context.Context, *tftypes.AttributePath, attr.Value) error
```

Reply:
Fixed in be40423bbd992e667de26a69e9e97b726a2abc8a